### ISSUE TITLE
Gone tasks gone are not cleaned-up correctly

### ISSUE THREAD

**Issue by benoit74** [openzim/zimfarm#877]
On athena18, we have a situation where:
- zimtask, zimscraper and dnstask containers are gone for some tasks
- tasks are still considered running in the Zimfarm
  - their status is not updated
  - the "workers" screen is still displaying them

Note that from an operational point of view, the impact is limited because the manager still sees that those tasks are gone and new tasks are hence requested and available resources are updated correctly.

Details found in Zimfarm Backend:
- task 43fd7183-f384-4bb1-9e20-71fda4dca53e: last communication is a Task Container ping at 
2023-11-25T06:19:36.803475943Z
- task 879b690e-8063-427d-ade5-b52c5617f233: last communication is a Task Container ping at 2023-11-25T06:19:19.636440088Z
- task f0c3a2b1-092b-4ca8-acc7-79edb3470edb: last communication is a Task Container ping at 2023-11-25T06:31:08.668670412Z
- task 61d848e1-7d5a-4b73-9aa3-e622bc963218:  last communication is a Task Container ping at 2023-11-25T06:31:39.472616998Z

`last reboot` shows that athena18 has been restarted at Sat Nov 25 18:09, so this is most probably linked.

The Zimfarm should nevertheless clean-up automatically the status of these tasks which are not running anymore on the worker, and most probably indicate a new status like "scraper_gone", clearly indicating that the scraper is probably not responsible for the task failure ; the corresponding recipes could even be requested again automatically or a clear alarm message could be sent to content team to warn them of the situation.

**Comment by benoit74** [comment]
> This is in place ; we should look after a regression here

Where ? The periodic cleanup is not looking after such situations AFAIK

**Comment by benoit74** [comment]
The `updated_at` task field is not even updated at every container ping, and the manager is not reporting currently running tasks from his PoV, so I do not see how this could be implemented.

**Comment by benoit74** [comment]
If I'm not mistaken, I think we can separate long-running tasks from a gone task, and continue to maintain the independence between manager, worker and API.

Currently:
- the worker task container is sending "Container pings" every minute ; we do not store the time this occurred (`updated_at` is not refreshed) but we could (we are updating the task to store container logs anyway)
- the worker manager is sending messages (requested-tasks) which in turn updates the `last_seen` property (at least it should if yesterday bug if fixed correctly :D)

Should the API stop to receive these worker task ping but still have the worker `last_seen` property updated, then we know the task should be cancelled (for safety, I would do a "cancel_requested", should there be any leftover containers, the manager should try to kill them) and marked as gone (probably a new field or we might use a specific "canceled_by" value)

**Comment by rgaudin** [comment]
But should the API but unreachable for some reason (API down or network issue between worker and API), then the API will mark it as gone and as soon as communication is restored, a cancel_request will be emitted…

**Comment by benoit74** [comment]
If API is unreachable, then the `last_seen` property won't be updated so the task won't be marked as gone. Or it means we would have an issue only between worker task and the API, but not between worker manager and the API. I don't think this would happen a lot.

We just have to specify reasonable delays to avoid race conditions (e.g. mark the task as gone only if it has not received a container ping for more than 10 minutes and the worker has been seen since less than 10 minutes ; 10 minutes or even 1 hour, we do not mind to let the system recover from temporary failures, we just do not want task to continue to be there for hours and days).

**Comment by rgaudin** [comment]
Sounds acceptable 👍

**Comment by benoit74** [comment]
If I resume the discussions, we will:
- update the `updated_at` task field at every container ping
- modify the `periodic_task.py` to check for gone tasks:
  - define a `STALLED_GONE_TIMEOUT` constant of one hour
  - modify `staled_statuses` to check for tasks which have an `updated_at` which is more than `STALLED_GONE_TIMEOUT` in the past and whose worker has an `last_seen` of less than `STALLED_GONE_TIMEOUT` in the past
    - for tasks matching these criteria, call the `task_cancel_requested_event_handler` with a proper payload

To have a clear understanding of these tasks which have been considered as gone, I see 2 options.

Option 1: set `canceled_by` set to `periodic-task-gone`
Option 2: add a `cancel_reason` field to tasks, add this somewhere in the UI, request this comment (mandatory) in the UI / API for user cancel requests

I clearly prefer option 2 even if more complex to develop, option 1 is too dirty and option 2 would benefit the system as a whole (if we agree on option 2, we should probably open a second issue to discuss this in more details, especially UI changes)

**Comment by rgaudin** [comment]
I don't think option 1 is too dirty. It serves the purpose of this ticket and in the context of this ticket, option2 brings no value.

I do agree that option 2 is better in general, especially given we've long wanted to have a log of ZF actions with the person taking an action and a mandatory reason.

If it were me, I'd do option 1 and redirect resources towards the log feature that is both a lot more work and a lot more useful (cancel-reason is hardly a question AFAIK). But you and @kelson42 are in charge of evolving the ZF so do as you see fit.

**Comment by rgaudin** [comment]
Yes that one. It's not in the ticket because almost all the UI allows is schedule modification but what you discussed here (task-cancellation) is one exception (but UI should probably be harmonized for both) and user management (but usage is exceptional)

**Comment by kelson42** [comment]
I have put a `prio 1` on this ticket because we remark that:
* With the multiplication of workers, such a scenario is not rare anymore
* Handling this manually has proven to be time consuming and difficult to do in a rigorous manner
* This is something which can and should be automatized.
* => Consequently we loose a significant amount of worker resources for almost no reason

@benoit74 How much effort would be needed to implement this ticket?

**Comment by kelson42** [comment]
> I don't think option 1 is too dirty. It serves the purpose of this ticket and in the context of this ticket, option2 brings no value.

Agree

> I do agree that option 2 is better in general, especially given we've long wanted to have a log of ZF actions with the person taking an action and a mandatory reason.

Yes, this is https://github.com/openzim/zimfarm/issues/498

> If it were me, I'd do option 1 and redirect resources towards the log feature that is both a lot more work and a lot more useful (cancel-reason is hardly a question AFAIK). But you and @kelson42 are in charge of evolving the ZF so do as you see fit.

Agree

**Comment by benoit74** [comment]
I think this would take at most 1 day of effort, the final design is very reasonable given we chose option 1 in the "proposed solution".

**Comment by benoit74** [comment]
Removed from zimit2 project unfortunately

**Comment by benoit74** [comment]
@elfkuzco this issue seems to be mostly ready to be developed.

Please read all comments to gain a good understanding of the situation, but to summarize them the idea is that you need to perform following modifications:
- update the `updated_at` task field at every container ping (call to `task_scraper_running_event_handler`)
- modify the `periodic_task.py` to check for gone tasks:
  - define a `STALLED_GONE_TIMEOUT` constant of one hour
  - modify `stalled_statuses` to check for tasks which have an `updated_at` which is more than `STALLED_GONE_TIMEOUT` in the past and whose worker has an `last_seen` of less than `STALLED_GONE_TIMEOUT` in the past
    - for tasks matching these criteria, call the `task_cancel_requested_event_handler` with a proper payload to cancel the task and set `canceled_by` to `periodic-task-gone`

The only concern is that if we continuously modify the `updated_at` task field at every container ping (more or less every minute or so), it looks like we will mess with sorting in quite a significant number of UI views, do you confirm? Do you have a solution to propose?

**Comment by elfkuzco** [comment]
> The only concern is that if we continuously modify the updated_at task field at every container ping (more or less every minute or so), it looks like we will mess with sorting in quite a significant number of UI views, do you confirm? Do you have a solution to propose?

There's not much issues with the UI in terms of `updated_at` being set on `scraper_running` events. Just that it would be more "live" in the sense that rows would be changing frequently.  Main discrepancy is that the `timestamp` for an event in the `events` list would be different from the `updated_at` for the task. For example, a task might have an event `scraper_completed` with a different `timestamp` from it's root `updated_at`.

**Comment by benoit74** [comment]
Ready in `zimfarm-upgrade` branch
